helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/ConstraintBasedAlgorithm.java

```diff
@@ -152,69 +179,147 @@ private double getAssignmentNormalizedScore(AssignableNode node, AssignableRepli
-            } else {
-              // In case of hash collision, return order according to the name.
-              return replica1.toString().compareTo(replica2.toString());
-            }
-          } else {
-            // Note we shall prioritize the replica with a higher state priority,
-            // the smaller priority number means higher priority.
-            return statePriority1 - statePriority2;
-          }
-        } else {
-          // If the baseline assignment contains the assignment, prioritize the replica.
-          return baselineAssignment.containsKey(resourceName1) ? -1 : 1;
+  private Pair<AssignableReplica, List<AssignableReplica>> getNextAssignableReplica(
+      List<AssignableReplica> allReplica,
+      Map<String, MutablePair<Integer, Integer>> overallClusterUtilMap, ClusterModel clusterModel,
+      Map<String, Integer> _replicaHashCodeMap) {
+    List<AssignableReplica> replicaForNextRound = new ArrayList<>();
+    float highestScore = -1;
+    AssignableReplica nextAssinableReplica = null;
+    Map<String, Float> weightMap = new HashMap<>();
+    // compute weight for each capacity dimension
+    for (Map.Entry<String, MutablePair<Integer, Integer>> resourceUsage : overallClusterUtilMap
+        .entrySet()) {
+      float freePercentage =
+          (float) (resourceUsage.getValue().right - resourceUsage.getValue().left) / resourceUsage
+              .getValue().right;
+      weightMap.put(resourceUsage.getKey(), freePercentage);
+    }
+    // Compare every replica with current candidate, update candidate if needed
+    for (AssignableReplica replica : allReplica) {
+      // score = SUM(weight * Resource_capacity) where weight = 1/(1-resource_util)
+      float score = 0.01f;
+      for (Map.Entry<String, Integer> resourceCapacity : replica.getCapacity().entrySet()) {
+        // use int to store % * 100
+        int replicaUsageInClusterPoolPct =
+            (resourceCapacity.getValue() * 100) / overallClusterUtilMap
+                .get(resourceCapacity.getKey()).right;
+        score += replicaUsageInClusterPoolPct / (weightMap.get(resourceCapacity.getKey()) + 0.01);
+      }
+      if (nextAssinableReplica == null) {
+        nextAssinableReplica = replica;
+        highestScore = score;
+      } else if (!keepOriginalPreferredReplica(nextAssinableReplica, replica, clusterModel,
+          _replicaHashCodeMap, highestScore, score)) {
+        highestScore = score;
+        if (nextAssinableReplica != null) {
+          replicaForNextRound.add(nextAssinableReplica);
         }
+        nextAssinableReplica = replica;
       } else {
-        // If the best possible assignment contains the assignment, prioritize the replica.
-        return bestPossibleAssignment.containsKey(resourceName1) ? -1 : 1;
+        replicaForNextRound.add(replica);
       }
-    });
-    return orderedAssignableReplicas;
+    }
+    return new MutablePair<>(nextAssinableReplica, replicaForNextRound);
```

Comment:
Why does `replicaForNextRound` need to be computed and returned by `getNextAssignableReplica`? It looks like it has no purpose in `getNextAssignableReplica`, because the logic is to find the replica with the highest score. 

Would it be more efficient to just remove the "best replica" from the `toBeAssignedReplicas` that's already computed in the outer scope?

Reply:
As discussed, the concern is that removing the "best replica" from the list is O(n). Please consider using a set for the outer scope data structure since the list-typing serves no purpose.

Reply:
Thanks for the review. I updated the PR based on our offline discussion.